### PR TITLE
Improve VariableWidthBlock deserialization

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Follows up from https://github.com/trinodb/trino/pull/15760 to further optimize `VariableWidthBlockEncoding#readBlock` to avoid allocating an unnecessary additional lengths array, since the final offsets array can temporarily store the serialized lengths of non-null positions and transform them back into valid offsets in-place.

Avoiding the extra array allocation does not reduce CPU throughput, and generally also improve as a result of being able to trigger CMOV optimizations a special path for the "no-nulls" case.

Benchmarks:
```
Benchmark                                  (nulls)  Mode  Before         After          Units 
BenchmarkBlockSerde.deserializeSliceDirect       0  avgt  4.174 ± 0.455  3.700 ± 0.092  ns/op
BenchmarkBlockSerde.deserializeSliceDirect     .01  avgt  4.943 ± 0.340  4.575 ± 0.330  ns/op
BenchmarkBlockSerde.deserializeSliceDirect     .10  avgt  5.471 ± 0.329  5.252 ± 0.078  ns/op
BenchmarkBlockSerde.deserializeSliceDirect     .50  avgt  6.730 ± 0.815  3.592 ± 0.167  ns/op <- CMOV
BenchmarkBlockSerde.deserializeSliceDirect     .90  avgt  2.636 ± 0.265  2.646 ± 0.260  ns/op
BenchmarkBlockSerde.deserializeSliceDirect     .99  avgt  1.303 ± 0.068  1.708 ± 0.038  ns/op
```

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java [hunk 1]
```diff
@@ -71,26 +71,52 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
...
+            }
+            // Simplified loop for no nulls present
+            for (int i = 1; i < offsets.length; i++) {
+                offsets[i] += offsets[i - 1];
+            }
+        }
+        else {
+            int currentOffset = 0;
+            int i = 1;
+            for (; i < offsets.length && lengthIndex < offsets.length; i++) {
```
Comment (radek-starburst) [inline core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java:97]: Why you dont initialize `i` in the `for` loop?

**Review by radek-starburst — APPROVED** [review]
LGTM but please run microbenchmarks

File: core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java [hunk 2]
```diff
@@ -71,26 +71,52 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
...
+                boolean isNull = valueIsNull[i - 1];
+                // must be accessed unconditionally, otherwise CMOV optimization isn't applied due to
+                // ArrayIndexOutOfBoundsException checks
+                int length = offsets[lengthIndex];
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java:107]: Benchmarks are included above in the PR description, the before / after columns are next to each other on each row.

[the remark this answers is not in the document]
Comment (radek-starburst) [reply core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java:107]: ok - cool. 
nit: please add it as commit description.

[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java:107]: Sure, done

**Comment by pettyjamesm** [comment]
@raunaqmorarka - let me know if you have any questions when you get a chance to review this PR.

**Comment by raunaqmorarka** [comment]
I haven't been able to reproduce improvements from this PR on my laptop
```
# JMH version: 1.35
# VM version: JDK 17.0.4, Java HotSpot(TM) 64-Bit Server VM, 17.0.4+11-LTS-179

Benchmark                                   (nullChance)  Mode  Cnt   Before          After          Units
BenchmarkBlockSerde.deserializeLineitem              N/A  avgt   60   3.830 ± 0.323   4.279 ± 0.315  ns/op
BenchmarkBlockSerde.deserializeSliceDirect             0  avgt   60   4.935 ± 0.246   4.530 ± 0.132  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .01  avgt   60   7.721 ± 0.308   7.653 ± 0.520  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .10  avgt   60   7.725 ± 0.227  13.047 ± 0.404  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .50  avgt   60  10.297 ± 0.175  10.485 ± 0.251  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .90  avgt   60   3.438 ± 0.105   7.606 ± 0.131  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .99  avgt   60   2.309 ± 0.168   5.180 ± 0.104  ns/op
```

```
Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro16,1
      Processor Name: 8-Core Intel Core i9
      Processor Speed: 2.4 GHz
      Number of Processors: 1
      Total Number of Cores: 8
      L2 Cache (per Core): 256 KB
      L3 Cache: 16 MB
      Hyper-Threading Technology: Enabled
      Memory: 64 GB
```

@radek-starburst could you please run this and post results from your system as well ?

**Comment by radek-starburst** [comment]
My results:

`                                                                                                         Before                 After
BenchmarkBlockSerde.deserializeSliceDirect             0  avgt   10  3.550 ± 0.296  ns/op   3.188 ± 0.125  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .01  avgt   10  4.860 ± 0.105  ns/op   4.655 ± 0.188  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .10  avgt   10  5.443 ± 0.162  ns/op   4.991 ± 0.161  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .50  avgt   10  6.312 ± 0.295  ns/op   3.530 ± 0.029  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .90  avgt   10  2.482 ± 0.098  ns/op   2.522 ± 0.054  ns/op
BenchmarkBlockSerde.deserializeSliceDirect           .99  avgt   10  1.288 ± 0.036  ns/op   1.618 ± 0.029  ns/op


`

File: core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java [hunk 3]
```diff
@@ -71,26 +73,51 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
...
+                lengthIndex = isNull ? lengthIndex : lengthIndex + 1;
+                currentOffset = isNull ? currentOffset : currentOffset + length;
+                offsets[i] = currentOffset;
+            }
+            if (lengthIndex != offsets.length) {
+                throw new IllegalArgumentException("Invalid nonNullsCount");
```
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java:113]: Please print the values suspected to be invalid in the exception message

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java (hunk 3 above)
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java]: Done

**Pull request merged**